broadinstitute/single_cell_portal_core
# Cell filtering by numeric annotations: basic UI (SCP-5385)

This enables users to filter cells by quantitative dimensions, and assess distributions in such annotations at a glance.

### Overview
Cell annotations can be categorical or numeric.  The cell filtering UI currently supports categorical (i.e. group-based) annotations, where users check and uncheck boxes for each group in an annotation.

Now, users can filter by numeric annotations as well.  This helps users easily answer important questions, for example "what was the level of this biomarker" or "what is this gene's expression in samples collected very soon after this experimental milestone".  The new basic UI for numeric cell filtering builds on recent library extensions (#1972).

### Video
Here's a brief demo!

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/9ec8fe87-36a4-4cfb-9057-c3a3fb0ae535

### Limitations
This UI for cell filtering by numeric annotations has basic functionality, but lacks the polish and richness of categorical cell filtering.  Given that, and to enable early feedback and easy wider testing, this new UI is behind a feature flag for now.  I plan to implement further refinements in my next PR.  Specific limitations and next steps for this feature are tracked in SCP-5500.

### Test
A new automated test confirms basic behavior, and existing tests are updated for this new functionality.  To manually test:

1.  If testing locally, run `rails db:migrate` in Rails console.  (If testing on staging, this was already done automatically.)
2.  Sign in, and enable the "show_numeric_cell_filtering" feature flag.  To do this, after signing in, open the upper-right menu and click "Feature flags".  Then select "user" in "Object Class", lookup by email, and find yourself.  Then select "True" for "show_numeric_cell_filtering" and click "Save".
3.  Go to a study with numeric annotations, e.g. "Cellular and transcriptional diversity over the course of human lactation".
4.  Click "Filter plotted cells".
5.  Find a numeric annotation, e.g. "BMI during study".
6.  In the input at left in this facet, enter a number slightly higher than the minimum.
7.  Confirm the scatter plot legend shows a "Filtered" entry.

This satisfies SCP-5385.
